### DISCUSSION TITLE
Why async_session does not provide .query method?

### DISCUSSION THREAD

**Discussion by mockingjet** [sqlalchemy/sqlalchemy#6564] [Usage Questions] [answered]
```py
class Book(Base):
    __tablename__ = "books"

    id = Column(Integer, autoincrement=True, primary_key=True, index=True)
    author = Column(String)


def get_books(session: Session, author):
    query = session.query(Book).filter(Book.author == author)
    return query.all()


async def get_books_async(session: AsyncSession, author):
    query = select(Book).where(Book.author == author)
    result = await session.execute(query)
    return result.scalars().all()
```

Why assync_session doesn't provide ".query" as session does?

I think it is much more convenient.

**Comment by CaselIT** [comment, marked as answer]
Hi,

Query is considered legacy and part of the v1 syntax of SQLAlchemy that it's not the suggested syntax to use when using the v2 style.
The main advantage of the new syntax is that it's unified with the core syntax, removing a lot of different api between orm and core objects.

Asyncio only supports the v2 style, so the query was not ported.

**Reply by mockingjet** [reply]
Okay. I understand it's for consistency with v2 style. Feel sad that "query" is not selected. 😢

**Reply by CaselIT** [reply]
One of the main goal of v2 was to reduce the api of sqlalchemy a bit, and one of the main duplication that exists is the select / query api that do more or less the same thing (allow construction of a select sql string) in two different ways.

So for v2 we tried to unify the api using select, that behaves a bit like both (example the select.join in 1.4 behaves the same as the query.join, not as it behaved in 1.3)

**Reply by mockingjet** [reply]
I do think session.query is more distinct for developers, not to mention it's no need to import select function.
Maybe session.query can be a sugar function replacing session.execute( ) + select

**Reply by zzzeek** [reply]
Background on Query becoming legacy is at: https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#orm-query-is-internally-unified-with-select-update-delete-2-0-style-execution-available

also it's not very hard to write a Query facade that does whatever methods you need, which creates a select() as it is used.  the sqlalchemy.orm.query.Query class itself can likely be subclassed directly to produce an asyncio version and I hope that third party libraries are introduced so that if people want to use this style they can, it's just no longer part of SQLAlchemy.

**Reply by ghevge** [reply]
The problem with dropping the support for legacy query is that for old code, it's a pain in the ass to convert tons of existing queries to new format. I am in a such situation. I'm trying to migrate to async_session an old tornado project. I have at least 500+ queries to update.... So I vote for adding legacy query support on async_session if possible. Thanks

**Reply by zzzeek** [reply]
I don't totally understand this.   SQLAlchemy's Query object does not offer "async/await" at all.    So by definition, if your old code uses Query, it does **not** use async/await meaning you have to convert all of your queries to a dramatically new format in any case.  Certainly converting an all-sync application to all-async is much more of the burden of effort rather than a slight change to how queries are started, for which you can easily write a wrapper function to do what you need?

**Reply by CaselIT** [reply]
Also note the existence of `run_sync` in the session and connection objects

**Reply by ghevge** [reply]
I was hoping to just put an await in from of my existing queries and job done. You guys don't make our life easy :)

**Reply by zzzeek** [reply]
what is async gaining you in that case if you dont already have lots of other async paradigms mixed in?   note that async doesn't create speed.

**Reply by ghevge** [reply]
The reason why I try this migration is because I currently have a random query hanging issue (in a tornado / sqlalchemy 1.3.x implementation) and I was hoping to solve that problem by  switching to async session.
